go-controller/pkg/util/pod_annotation.go

```diff
@@ -371,3 +376,330 @@ func GetK8sPodAllNetworkSelections(pod *v1.Pod) ([]*nadapi.NetworkSelectionEleme
+			return nil
+		}
+
+		err = kube.UpdatePod(pod)
+		if err != nil && rollback != nil {
+			rollback()
+		}
+
+		return err
+	})
+
+	if err != nil {
+		return fmt.Errorf("failed to update pod %s/%s: %w", pod.Namespace, pod.Name, err)
+	}
+
+	return nil
+}
+
+// UpdatePodAnnotationWithRetry updates the pod annotation on the pod retrying
+// on conflict
+func UpdatePodAnnotationWithRetry(podLister listers.PodLister, kube kube.Interface, pod *v1.Pod, podAnnotation *PodAnnotation, nadName string) error {
+	updatePodAnnotationNoRollback := func(pod *v1.Pod) (*v1.Pod, func(), error) {
+		var err error
+		pod.Annotations, err = MarshalPodAnnotation(pod.Annotations, podAnnotation, nadName)
+		if err != nil {
+			return nil, nil, err
+		}
+		return pod, nil, nil
+	}
+
+	return UpdatePodWithAllocationOrRollback(
+		podLister,
+		kube,
+		pod,
+		updatePodAnnotationNoRollback,
+	)
+}
+
+// ipAllocator is an IP allocation interface that needs to be implemented to use
+// AllocatePodAnnotationWithRollback
+type ipAllocator interface {
+	AllocateIPs(ips []*net.IPNet) error
+	AllocateNextIPs() ([]*net.IPNet, error)
+	ReleaseIPs(ips []*net.IPNet) error
+	IsErrAllocated(error) bool
+}
+
+// AllocatePodAnnotationWithRollback allocates the pod annotation which includes
+// IPs, a mac address and gateways. The allocation can be requested through the
+// network selection element or derived from the allocator provided IPs. If the
+// requested IPs cannot be honored, a new set of IPs will be allocated unless
+// reallocate is set to false. A rollback function is returned to rollback the
+// IP allocation if there was one. This function is designed to be used in
+// AllocateToPodWithRollbackFunc implementations. Use an inlined implementation
+// if you want to extract information from it as a side-effect.
+func AllocatePodAnnotationWithRollback(
+	netInfo NetInfo,
+	pod *v1.Pod,
+	podAnnotation *PodAnnotation,
+	network *nadapi.NetworkSelectionElement,
+	ipAllocator ipAllocator,
+	reallocate bool) (
+	updatedPod *v1.Pod,
+	updatedPodAnnotation *PodAnnotation,
+	rollback func(), err error) {
+
+	nadName := types.DefaultNetworkName
+	if netInfo.IsSecondary() {
+		nadName = GetNADName(network.Namespace, network.Name)
+	}
+	podDesc := fmt.Sprintf("%s/%s/%s", nadName, pod.Namespace, pod.Name)
+
+	// the IPs we allocate in this function need to be released back to the IPAM
+	// pool if there is some error in any step past the point the IPs were
+	// assigned via the IPAM manager. Note we are using a named return variable
+	// for defer to work correctly.
+	var releaseIPs []*net.IPNet
+	rollback = func() {
+		if len(releaseIPs) == 0 {
+			return
+		}
+		err := ipAllocator.ReleaseIPs(releaseIPs)
+		if err != nil {
+			klog.Errorf("Error when releasing IPs %v: %w", StringSlice(releaseIPs), err)
+			releaseIPs = nil
+			return
+		}
+		releaseIPs = nil
+		klog.Infof("Released IPs %v", StringSlice(releaseIPs))
+	}
+	defer func() {
+		if err != nil {
+			rollback()
```

Comment:
We return a rollback function to allow the possibility to rollback the allocation to this function user even when there was no error within. It is used in `AllocateToPodWithRollbackFunc` and explained in the function doc. Let me know how this could be clarified further.

Reply:
My question was misleading, sorry for that.

I meant to ask if we shouldn't return `nil` as rollback function when there **is** an error. All in all, you've already rolled it back. I guess it is up to the caller to never roll it back when it sees an error then.

You're good, as you said, your docstring explains quite well what this is supposed to do.

Reply:
Yeah, we could do that. But we do return something that can be used but does nothing so that callers are not forced to check for nil, which they should but I don't see that as ingrained in golang as let's say C. I thought there's no harm in returning a noop rollback.

Reply:
totally fine. 

Maybe the docstring should say the rollback functions must be idempotent ?